### PR TITLE
Add gshard implementation w/ balancing loss, etc.

### PR DESCRIPTION
**Pull request by hadasah**

### CODE REVIEW

File: fairseq/criterions/cross_entropy.py [hunk 1]
```diff
@@ -88,3 +88,79 @@ def logging_outputs_can_be_summed() -> bool:
...
+        lprobs = lprobs.view(-1, lprobs.size(-1))
+        target = model.get_targets(sample, net_output).view(-1)
+        loss = F.nll_loss(
+            lprobs,
+            target,
+            ignore_index=self.padding_idx,
+            reduction="sum" if reduce else "none",
+        )
+        loss += gate_loss
+        return loss, loss
```
Comment (TimDettmers) [inline fairseq/criterions/cross_entropy.py]: It is better to return `loss+gate_loss, loss` and remove `loss += gate_loss`. This way the evaluation and train perplexity that is logged does not contain the balancing loss term.

File: fairseq/models/transformer/transformer_decoder.py [hunk 2]
```diff
@@ -114,12 +116,52 @@ def __init__(
             self.layers = LayerDropModuleList(p=self.decoder_layerdrop)
         else:
             self.layers = nn.ModuleList([])
+
+        num_moe_layers = cfg.moe_layers
+        moe_indices = cfg.moe_layer_indices
+        extra_moe_indices = []
+        if moe_indices is None:
+            # if no indices were specified, intersperse them as in the original papers
+            moe_indices = [((i + 1) * cfg.decoder.layers) // (num_moe_layers + 1) for i in range(num_moe_layers)]
```
Comment (TimDettmers) [inline fairseq/models/transformer/transformer_decoder.py:124]: It could make sense to ensure that the first transformer block is dense. Not sure if this is like that in the original papers, but my impression was that MoE layers need some solid features to begin with, so I would make sure that the first layer is always dense.

File: fairseq/models/transformer/transformer_decoder.py [hunk 3]
```diff
@@ -114,12 +116,52 @@ def __init__(
...
+                #     cfg, shared_moe_layer=shared_moe_layer, shared_moe_experts=shared_moe_experts
+                # ) if i in moe_indices
+                self.build_decoder_layer(
+                    cfg, no_encoder_attn=no_encoder_attn, shared_moe_layer=shared_moe_layer, 
+                    shared_moe_experts=shared_moe_experts, moe=(True if i in moe_indices else False)
+                ) for i in range(cfg.decoder.layers)
             ]
         )
+
+        if extra_moe_indices:
```
Comment (TimDettmers) [inline fairseq/models/transformer/transformer_decoder.py:152]: This is currently empty but could be used to have a mix of shared and non-shared expert layers -- is that right?

File: fairseq/models/transformer/transformer_decoder.py [hunk 4]
```diff
@@ -163,15 +205,19 @@ def build_output_projection(self, cfg, dictionary, embed_tokens):
...
-                BaseLayer(cfg),
-            )
 
-    def build_decoder_layer(self, cfg, no_encoder_attn=False):
-        layer = transformer_layer.TransformerDecoderLayerBase(cfg, no_encoder_attn)
+    def build_decoder_layer(
+        self, cfg, no_encoder_attn=False, 
+        shared_moe_layer=None, shared_moe_experts=None, moe=False,
+    ):
+        # TODO @margsli does this need to be modified for MoE?
```
Comment (TimDettmers) [inline fairseq/models/transformer/transformer_decoder.py]: Anything, in particular, you have in mind? I think this should be okay.

File: fairseq/modules/moe_layer.py [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+
+        if self.shuffle and is_training:
+            # Send each token to a random worker, to break correlations within the batch
+            shuffle_sort = torch.randperm(features.size(0), device=features.device)
+            features = All2All.apply(features[shuffle_sort])
+
+        with torch.no_grad():
+            # Compute similarity of each token to each expert, for routing
+            token_expert_affinities = features.matmul(self.expert_centroids.transpose(0, 1))
+            # TODO @margsli should be identical to below, is the einsum faster?
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py]: Einsum might take a different path through computation operations, but it should not be much faster. You can give it a try and look at the words per seconds logged by fairseq for a couple of updates and then just keep what is faster. My guess it will be the same speed.

File: fairseq/modules/moe_layer.py [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+from torch import Tensor
+
+
+class MoELayer(nn.Module):
+
+    def __init__(self, args, shared_moe_experts=None):
+        super().__init__()
+        self.num_workers = distributed_utils.get_data_parallel_world_size()
+        expert_centroids = torch.empty(self.num_workers, args.decoder_embed_dim)
+        torch.nn.init.orthogonal_(expert_centroids, gain=0.1)
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py]: I have never used orthogonal init before. I am curious why you chose this

File: fairseq/modules/moe_layer.py [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+        # Creates an index that undoes a sort: xs==xs[order][inverse_sort(order)]
+        return torch.empty_like(order).scatter_(0, order, torch.arange(0, order.size(0), device=order.device))
+
+    # Assigns each token to the top k experts
+    # NOTE: this ONLY works for k=1 right now b/c of self.last_mask assignment
+    def assignment(self, scores, strategy='greedy', k=1):
+        if strategy == 'greedy':
+            # get a flattened list of the workers to send each token to
+            # scores is s,b,e, token_to_workers is s*b, with the worker # for each token
+            idx = torch.topk(scores, dim=1, k=k, largest=True).indices
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py]: If scores is s,b,e should dim not be `dim=2`?

File: fairseq/modules/moe_layer.py [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+            self.last_mask = F.one_hot(idx[:, 0], num_classes=self.num_workers)
+            token_to_workers = idx.view(-1)
+            # token_to_workers is worker # for each token, sorted, sort_ordering gives original index
+            token_to_workers, sort_ordering = torch.sort(token_to_workers)
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py:127]: Just to understand what is going on here: You have dimensions of `[b, s]` and now you sort the entire array, that gives you the sorted indices (consecutive tokens to send to each expert) and the indices for that. 

Then you do indices // k to distribute them to workers in the case you have multiple experts per GPU?

File: fairseq/modules/moe_layer.py [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+            # this works because 2n and 2n+1 are the same token, diff experts
+            worker2token = sort_ordering // k
+
+            # Find how many tokens we're sending to each other worker (being careful for sending 0 tokens to some workers)
+            output_splits = torch.zeros((self.num_workers,), dtype=torch.long, device=scores.device)
+            workers, counts = torch.unique_consecutive(token_to_workers, return_counts=True)
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py:133]: This is pretty nifty! Have not thought about doing it like that

File: fairseq/modules/moe_layer.py [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+
+        # Compute which token goes to which expert
+        routing_strategy = 'greedy'
+        sort_by_expert, input_splits, output_splits = self.assignment(token_expert_affinities, strategy=routing_strategy)
+        # Swap these tokens for the right ones for our expert
+        routed_features = All2All.apply(features[sort_by_expert], output_splits, input_splits)
+
+        if routed_features.size(0) > 0:
+            # Mix in the expert network based on how appropriate it is for these tokens
+            alpha = torch.sigmoid(routed_features.mv(self.expert_centroids[self.expert_id])).unsqueeze(1)
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py]: Why the logistic sigmoid here? What does the `(1- alpha)` do here?

File: fairseq/modules/moe_layer.py [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+            routed_features = alpha * self.expert_network(routed_features) + (1 - alpha) * routed_features
+        # Return to original worker and ordering
+        result = All2All.apply(routed_features, input_splits, output_splits)[self.inverse_sort(sort_by_expert)]
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py:73]: Can you run me through the logic behind this? I cannot quite follow if there is any difference in terms of order for (1) All2All and (2) inverse_sort.

Comment (TimDettmers) [reply fairseq/modules/moe_layer.py:73]: I think I get it now. You reverse the input and output splits and that reverses the indices and should yield the correct all2all mapping. Afterward, you reorder the sorted all2all output to its original order. Is that correct?

File: fairseq/modules/moe_layer.py [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+        if self.shuffle and is_training:
+            # Undo shuffling
+            result = All2All.apply(result)[self.inverse_sort(shuffle_sort)]
+
+        # Return additional Nones for compatibility with TransformerDecoderLayer
+        return result.view(input_features.size()), None, None
+
+    def inverse_sort(self, order):
+        # Creates an index that undoes a sort: xs==xs[order][inverse_sort(order)]
+        return torch.empty_like(order).scatter_(0, order, torch.arange(0, order.size(0), device=order.device))
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py:84]: It is easy to introduce an error here and the network still works. I find it difficult to check it mentally. Maybe we can talk about this in the meeting and run through the dimensions etc

File: fairseq/modules/moe_layer.py [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+        self.cross_self_attention = cfg.cross_self_attention
+
+        self.self_attn = self.build_self_attention(
+            self.embed_dim,
+            cfg,
+            add_bias_kv=add_bias_kv,
+            add_zero_attn=add_zero_attn,
+        )
+
+        # TODO @margsli Does it matter if I use the simpler relu activation code above or use this?
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py:259]: Looks to me as almost the same. I think it is just a matter of style and readability, but both seem fine to me

File: fairseq/modules/moe_layer.py [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+        # if self.normalize_before:
+        #     x = self.final_layer_norm(x)
+
+        # x = self.activation_fn(self.fc1(x))
+        # x = self.activation_dropout_module(x)
+        # x = self.fc2(x)
+        x = self.moe(x)
+
+        x = self.dropout_module(x)
+        x = self.residual_connection(x, residual)
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py:460]: I think this residual currently take the input into the block but the residual should be the input into the moe layer

File: fairseq/modules/moe_layer.py [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+                need_head_weights=need_head_weights,
+            )
+            x = self.dropout_module(x)
+            x = self.residual_connection(x, residual)
+            if not self.normalize_before:
+                x = self.encoder_attn_layer_norm(x)
+
+        residual = x
+        # if self.normalize_before:
+        #     x = self.final_layer_norm(x)
```
Comment (TimDettmers) [inline fairseq/modules/moe_layer.py]: Not quite sure if the layer norms are in the right place. I think you just moved them around, but not fully sure. Make sure you have the layer norms in the place you want them to be

**Review by TimDettmers — COMMENTED** [review]
This looks pretty good! I like the code -- very easy to read. Using the consecutive sequences method is pretty neat too.

I think there were only a couple of issues that confused me. Possibly one error in the topk `dim` variable. Otherwise, there are a couple of things for you to check that I do not fully understand if they are correct, but it's probably easy for you to check for correctness. Otherwise only a couple of things that we need to discuss further in our meeting, but happy to discuss anything really. You can post some comments here too if that is an easier way to discuss

File: fairseq/models/transformer/transformer_decoder.py (hunk 2 above)
Comment (TimDettmers) [reply fairseq/models/transformer/transformer_decoder.py:124]: Actually not sure if this code already does this. I think right now the first index is 0, but could be mistaken

File: fairseq/criterions/cross_entropy.py (hunk 1 above)
Comment (hadasah) [reply fairseq/criterions/cross_entropy.py]: I actually just this weekend added in fairseq's logging so I ended up splitting this up :)

File: fairseq/models/transformer/transformer_decoder.py (hunk 2 above)
Comment (hadasah) [reply fairseq/models/transformer/transformer_decoder.py:124]: if num_moe_layers == num_decoder_layers then of course all the layers are dense, but otherwise I think the 0th layer is always dense. But I can add code to throw an error if the first layer isn't dense

File: fairseq/models/transformer/transformer_decoder.py (hunk 3 above)
Comment (hadasah) [reply fairseq/models/transformer/transformer_decoder.py:152]: yes, after this extend call, self.layers should contain all of the decoder layers, and if the 0th and 3rd layers are shared, for instance, then self.layers[0] == self.layers[3]. I'm not sure if this is the correct/best way to implement the weight sharing, but it was the simplest thing to do.

File: fairseq/modules/moe_layer.py (hunk 5 above)
Comment (hadasah) [reply fairseq/modules/moe_layer.py]: 👍

File: fairseq/modules/moe_layer.py (hunk 7 above)
Comment (hadasah) [reply fairseq/modules/moe_layer.py]: yes, I was confused about this when I wrote it, but it turns out scores has been flattened above to sxb, e, so this is actually correct -- I was going to change it to dim=-1 originally

File: fairseq/modules/moe_layer.py (hunk 9 above)
Comment (hadasah) [reply fairseq/modules/moe_layer.py:133]: this is all Mike haha I thought it was very clever too!! But it took me a few tries to understand what he did.

File: fairseq/modules/moe_layer.py (hunk 8 above)
Comment (hadasah) [reply fairseq/modules/moe_layer.py:127]: yes, except I was actually wrong, and it's just `[bxs]`  -- as in, there's only one dimension. The input got flattened in line 48 above: in the forward pass of MoELayer with the line `features = input_features.reshape(-1, input_features.size(-1))`

File: fairseq/modules/moe_layer.py (hunk 14 above)
Comment (hadasah) [reply fairseq/modules/moe_layer.py:460]: In 410, there's `residual = x` which should be setting it to the input to the moe layer unless I'm mistaken? I'm not 100% sure

File: fairseq/modules/moe_layer.py (hunk 15 above)
Comment (hadasah) [reply fairseq/modules/moe_layer.py]: ah, I accidentally removed the final layernorm, which should happen right after the moe block. There is a layernorm right after the enc-dec attention, then it actually goes straight into *another* layernorm in the moe, since that's how base layers had it. I think the idea was that base layers was doing/wanted to do prenorm, but the default is postnorm. I don't know if there's a better or worse choice?

File: fairseq/modules/moe_layer.py (hunk 13 above)
Comment (hadasah) [reply fairseq/modules/moe_layer.py:259]: 👍

**Review by hadasah — COMMENTED** [review]
Thanks so much for the comments @TimDettmers! I made a mistake in the comments because I originally misunderstood the dimensions, so I'm going to update the branch with the fixed comments and some stuff I copied from fairseq's gshard/switch code (mostly just logging).
